src/reference.cr

```diff
@@ -135,8 +135,21 @@ class Reference
 
   # :nodoc:
   module ExecRecursive
+    alias Registry = Hash({UInt64, Symbol}, Bool)
+
+    {% if flag?(:preview_mt) %}
+      @@exec_recursive : Crystal::ThreadLocalValue(Registry)?
+    {% else %}
+      @@exec_recursive : Registry?
+    {% end %}
+
     def self.hash
-      @@exec_recursive ||= {} of {UInt64, Symbol} => Bool
+      {% if flag?(:preview_mt) %}
+        exec_recursive = @@exec_recursive ||= Crystal::ThreadLocalValue(Registry).new
```

Comment:
This initialization should be protected by a Mutex or moved as a class variable initializer. Otherwise a race condition will happen and multiple instances of `Crystal::ThreadLocalValue` are created.

Reply:
Fixed!